File: electron/ipc/filesystem.ts
import { ipcMain, shell } from 'electron';
import path from 'node:path';
import fs from 'node:fs';
import log from 'electron-log';
import { isPathWithinNotesDir } from '../utils/config';

let notesDir: string;

export function setNotesDir(dir: string): void {
  notesDir = dir;
}

export function openNotesFolder(): void {
  shell.openPath(notesDir);
}

export function registerFilesystemHandlers(): void {
  ipcMain.handle('delete-folder', async (_event, folderPath: string) => {
    try {
      if (!folderPath || typeof folderPath !== 'string') {
        return { success: false, error: 'Invalid folder path' };
      }
      
      if (!isPathWithinNotesDir(folderPath, notesDir)) {
        log.error('Attempted to delete folder outside notes directory:', folderPath);
        return { success: false, error: 'Access denied: path outside notes directory' };
      }
      
      if (fs.existsSync(folderPath)) {
        fs.rmSync(folderPath, { recursive: true, force: true });
        return { success: true };
      }
      return { success: false, error: 'Folder not found' };
    } catch (e) {
      log.error('Error deleting folder:', e);
      return { success: false, error: String(e) };
    }
  });

  ipcMain.handle('has-md-files', async (_event, dirPath: string): Promise<boolean> => {
    if (!dirPath || typeof dirPath !== 'string') {
      return false;
    }
    
    const checkDir = (dir: string): boolean => {
      try {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
          if (entry.name.startsWith('.')) continue;
          const fullPath = path.join(dir, entry.name);
          if (entry.isDirectory()) {
            if (checkDir(fullPath)) return true;
          } else if (entry.name.endsWith('.md')) {
            return true;
          }
        }
      } catch {
        return false;
      }
      return false;
    };
    return checkDir(dirPath);
  });

  ipcMain.handle('create-folder', async (_event, parentPath: string, folderName: string) => {
    try {
      if (!parentPath || typeof parentPath !== 'string') {
        return { success: false, error: 'Invalid parent path' };
      }
      
      if (!folderName || typeof folderName !== 'string') {
        return { success: false, error: 'Invalid folder name' };
      }
      
      const sanitizedName = folderName.replace(/[^a-zA-Z0-9\-]/g, '-');
      
      if (!isPathWithinNotesDir(parentPath, notesDir)) {
        log.error('Attempted to create folder outside notes directory:', parentPath);
        return { success: false, error: 'Access denied' };
      }
      
      const newPath = path.join(parentPath, sanitizedName);
      
      if (fs.existsSync(newPath)) {
        return { success: false, error: 'A folder with this name already exists' };
      }
      
      fs.mkdirSync(newPath, { recursive: true });
      return { success: true, path: newPath };
    } catch (e) {
      log.error('Error creating folder:', e);
      return { success: false, error: String(e) };
    }
  });

  ipcMain.handle('rename-folder', async (_event, oldPath: string, newName: string) => {
    try {
      if (!oldPath || typeof oldPath !== 'string') {
        return { success: false, error: 'Invalid folder path' };
      }
      
      if (!newName || typeof newName !== 'string') {
        return { success: false, error: 'Invalid folder name' };
      }
      
      const sanitizedName = newName.replace(/[^a-zA-Z0-9\-]/g, '-');
      
      if (!isPathWithinNotesDir(oldPath, notesDir)) {
        log.error('Attempted to rename folder outside notes directory:', oldPath);
        return { success: false, error: 'Access denied' };
      }
      
      const dir = path.dirname(oldPath);
      const newPath = path.join(dir, sanitizedName);
      
      if (fs.existsSync(newPath) && newPath !== oldPath) {
        return { success: false, error: 'A folder with this name already exists' };
      }
      
      fs.renameSync(oldPath, newPath);
      return { success: true, newPath };
    } catch (e) {
      log.error('Error renaming folder:', e);
      return { success: false, error: String(e) };
    }
  });

  ipcMain.handle('move-file', async (_event, sourcePath: string, destFolder: string) => {
    try {
      if (!sourcePath || typeof sourcePath !== 'string') {
        return { success: false, error: 'Invalid source path' };
      }
      
      if (!destFolder || typeof destFolder !== 'string') {
        return { success: false, error: 'Invalid destination folder' };
      }
      
      if (!isPathWithinNotesDir(sourcePath, notesDir)) {
        log.error('Attempted to move file outside notes directory:', sourcePath);
        return { success: false, error: 'Access denied' };
      }
      
      if (!isPathWithinNotesDir(destFolder, notesDir)) {
        log.error('Attempted to move file to outside notes directory:', destFolder);
        return { success: false, error: 'Access denied' };
      }
      
      const fileName = path.basename(sourcePath);
      let destPath = path.join(destFolder, fileName);
      
      if (fs.existsSync(destPath)) {
        const nameWithoutExt = fileName.replace(/\.md$/, '');
        let counter = 1;
        let newFileName = `${nameWithoutExt}-${counter}.md`;
        destPath = path.join(destFolder, newFileName);
        
        while (fs.existsSync(destPath)) {
          counter++;
          newFileName = `${nameWithoutExt}-${counter}.md`;
          destPath = path.join(destFolder, newFileName);
        }
      }
      
      fs.renameSync(sourcePath, destPath);
      return { success: true, newPath: destPath };
    } catch (e) {
      log.error('Error moving file:', e);
      return { success: false, error: String(e) };
    }
  });

  ipcMain.handle('get-directory', async (_event, dirPath: string) => {
    try {
      if (!dirPath || typeof dirPath !== 'string') {
        return [];
      }
      
      const entries = fs.readdirSync(dirPath, { withFileTypes: true });
      return entries
        .filter(entry => !entry.name.startsWith('.') && (entry.isDirectory() || entry.name.endsWith('.md')))
        .map(entry => {
          const fullPath = path.join(dirPath, entry.name);
          return {
            name: entry.name,
            path: fullPath,
            isDirectory: entry.isDirectory()
          };
        })
        .sort((a, b) => {
          if (a.isDirectory !== b.isDirectory) {
            return a.isDirectory ? -1 : 1;
          }
          return a.name.localeCompare(b.name);
        });
    } catch (e) {
      log.error('Error reading directory:', e);
      return [];
    }
  });

  ipcMain.handle('read-file', async (_event, filePath: string) => {
    try {
      if (!filePath || typeof filePath !== 'string') {
        return null;
      }
      
      const content = fs.readFileSync(filePath, 'utf-8');
      const name = path.basename(filePath, '.md');
      const stats = fs.statSync(filePath);
      return {
        id: filePath,
        title: name,
        content: content,
        createdAt: stats.birthtimeMs,
        updatedAt: stats.mtimeMs
      };
    } catch (e) {
      log.error('Error reading file:', e);
      return null;
    }
  });
}

File: src-tauri/src/commands/filesystem.rs
use crate::commands::{FileEntry, OperationResult};
use std::fs;
use std::path::Path;
use tauri::State;
use walkdir::WalkDir;

use crate::commands::notes::AppState;

fn is_path_within_notes_dir(file_path: &str, notes_dir: &str) -> bool {
    let resolved_path = Path::new(file_path).canonicalize().unwrap_or_default();
    let resolved_notes_dir = Path::new(notes_dir).canonicalize().unwrap_or_default();
    resolved_path.starts_with(&resolved_notes_dir)
}

#[tauri::command]
pub fn delete_folder(state: State<'_, AppState>, folder_path: String) -> OperationResult {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&folder_path, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied: path outside notes directory".to_string()),
            new_path: None,
            path: None,
        };
    }

    if Path::new(&folder_path).exists() {
        match fs::remove_dir_all(&folder_path) {
            Ok(_) => OperationResult {
                success: true,
                error: None,
                new_path: None,
                path: None,
            },
            Err(e) => OperationResult {
                success: false,
                error: Some(e.to_string()),
                new_path: None,
                path: None,
            },
        }
    } else {
        OperationResult {
            success: false,
            error: Some("Folder not found".to_string()),
            new_path: None,
            path: None,
        }
    }
}

#[tauri::command]
pub fn has_md_files(_state: State<'_, AppState>, dir_path: String) -> bool {
    for entry in WalkDir::new(&dir_path)
        .follow_links(true)
        .max_depth(10)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();
        if path.is_file() && path.extension().map_or(false, |ext| ext == "md") {
            return true;
        }
    }
    false
}

#[tauri::command]
pub fn create_folder(
    state: State<'_, AppState>,
    parent_path: String,
    folder_name: String,
) -> OperationResult {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&parent_path, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied".to_string()),
            new_path: None,
            path: None,
        };
    }

    let sanitized_name = folder_name.replace(|c: char| !c.is_alphanumeric() && c != '-', "-");
    let new_path = Path::new(&parent_path).join(&sanitized_name);

    if new_path.exists() {
        return OperationResult {
            success: false,
            error: Some("A folder with this name already exists".to_string()),
            new_path: None,
            path: None,
        };
    }

    match fs::create_dir_all(&new_path) {
        Ok(_) => OperationResult {
            success: true,
            error: None,
            new_path: None,
            path: Some(new_path.to_string_lossy().to_string()),
        },
        Err(e) => OperationResult {
            success: false,
            error: Some(e.to_string()),
            new_path: None,
            path: None,
        },
    }
}

#[tauri::command]
pub fn rename_folder(
    state: State<'_, AppState>,
    old_path: String,
    new_name: String,
) -> OperationResult {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&old_path, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied".to_string()),
            new_path: None,
            path: None,
        };
    }

    let sanitized_name = new_name.replace(|c: char| !c.is_alphanumeric() && c != '-', "-");
    let parent = Path::new(&old_path).parent().unwrap_or(Path::new(&notes_dir));
    let new_path = parent.join(&sanitized_name);

    if new_path.exists() && new_path != Path::new(&old_path) {
        return OperationResult {
            success: false,
            error: Some("A folder with this name already exists".to_string()),
            new_path: None,
            path: None,
        };
    }

    match fs::rename(&old_path, &new_path) {
        Ok(_) => OperationResult {
            success: true,
            error: None,
            new_path: Some(new_path.to_string_lossy().to_string()),
            path: None,
        },
        Err(e) => OperationResult {
            success: false,
            error: Some(e.to_string()),
            new_path: None,
            path: None,
        },
    }
}

#[tauri::command]
pub fn move_file(
    state: State<'_, AppState>,
    source_path: String,
    dest_folder: String,
) -> OperationResult {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&source_path, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied".to_string()),
            new_path: None,
            path: None,
        };
    }

    if !is_path_within_notes_dir(&dest_folder, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied".to_string()),
            new_path: None,
            path: None,
        };
    }

    let file_name = Path::new(&source_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    let mut dest_path = Path::new(&dest_folder).join(&file_name);

    if dest_path.exists() {
        let name_without_ext = dest_path
            .file_stem()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();
        let mut counter = 1;
        let mut new_file_name = format!("{}-{}.md", name_without_ext, counter);
        dest_path = Path::new(&dest_folder).join(&new_file_name);

        while dest_path.exists() {
            counter += 1;
            new_file_name = format!("{}-{}.md", name_without_ext, counter);
            dest_path = Path::new(&dest_folder).join(&new_file_name);
        }
    }

    match fs::rename(&source_path, &dest_path) {
        Ok(_) => OperationResult {
            success: true,
            error: None,
            new_path: Some(dest_path.to_string_lossy().to_string()),
            path: None,
        },
        Err(e) => OperationResult {
            success: false,
            error: Some(e.to_string()),
            new_path: None,
            path: None,
        },
    }
}

#[tauri::command]
pub fn create_file(
    state: State<'_, AppState>,
    dir_path: String,
    file_name: String,
    content: Option<String>,
) -> OperationResult {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&dir_path, &notes_dir) {
        return OperationResult {
            success: false,
            error: Some("Access denied".to_string()),
            new_path: None,
            path: None,
        };
    }

    let sanitized_name = file_name.replace(|c: char| !c.is_alphanumeric() && c != '-', "-");
    let file_path = Path::new(&dir_path).join(format!("{}.md", sanitized_name));

    if file_path.exists() {
        return OperationResult {
            success: false,
            error: Some("A file with this name already exists".to_string()),
            new_path: None,
            path: None,
        };
    }

    let default_content = content.unwrap_or_else(|| {
        let date = chrono::Utc::now().format("%Y-%m-%d").to_string();
        format!(
            "---\ntitle: {}\ndate: {}\ntags: []\n#\n---\n",
            sanitized_name, date
        )
    });

    match fs::write(&file_path, default_content) {
        Ok(_) => OperationResult {
            success: true,
            error: None,
            new_path: None,
            path: Some(file_path.to_string_lossy().to_string()),
        },
        Err(e) => OperationResult {
            success: false,
            error: Some(e.to_string()),
            new_path: None,
            path: None,
        },
    }
}

#[tauri::command]
pub fn get_directory(dir_path: String) -> Vec<FileEntry> {
    let path = Path::new(&dir_path);
    if !path.exists() || !path.is_dir() {
        return Vec::new();
    }

    let mut entries: Vec<FileEntry> = Vec::new();

    if let Ok(read_dir) = fs::read_dir(path) {
        for entry in read_dir.filter_map(|e| e.ok()) {
            let entry_path = entry.path();
            let name = entry_path.file_name().unwrap_or_default().to_string_lossy().to_string();

            if name.starts_with('.') {
                continue;
            }

            let is_dir = entry_path.is_dir();
            if is_dir || name.ends_with(".md") {
                entries.push(FileEntry {
                    name,
                    path: entry_path.to_string_lossy().to_string(),
                    is_directory: is_dir,
                });
            }
        }
    }

    entries.sort_by(|a, b| {
        if a.is_directory != b.is_directory {
            b.is_directory.cmp(&a.is_directory)
        } else {
            a.name.to_lowercase().cmp(&b.name.to_lowercase())
        }
    });

    entries
}

use crate::commands::FileReadResult;

#[tauri::command]
pub fn read_file(state: State<'_, AppState>, file_path: String) -> Option<FileReadResult> {
    let notes_dir = state.notes_dir.lock().unwrap().clone();

    if !is_path_within_notes_dir(&file_path, &notes_dir) {
        log::error!("read_file: Access denied for path: {}", file_path);
        return None;
    }

    let path = Path::new(&file_path);
    if !path.exists() {
        log::error!("read_file: File not found: {}", file_path);
        return None;
    }

    let content = fs::read_to_string(path).ok()?;
    let name = path.file_stem().unwrap_or_default().to_string_lossy().to_string();
    let metadata = fs::metadata(path).ok();
    let props = crate::commands::parse_frontmatter_full(&content);
    let tags = crate::commands::parse_tags(&props);

    Some(FileReadResult {
        id: file_path,
        title: name,
        content,
        created_at: metadata
            .as_ref()
            .and_then(|m| m.created().ok())
            .map(|t| t.duration_since(std::time::UNIX_EPOCH).unwrap_or_default().as_millis() as u64)
            .unwrap_or(0),
        updated_at: metadata
            .as_ref()
            .and_then(|m| m.modified().ok())
            .map(|t| t.duration_since(std::time::UNIX_EPOCH).unwrap_or_default().as_millis() as u64)
            .unwrap_or(0),
        tags,
        properties: props,
    })
}
